src/schema.jl

```diff
@@ -0,0 +1,197 @@
+
+# TODO: handle streaming (Data.RowTable) by iterating over rows and updating
+# schemas in place
+
+terms(t::FormulaTerm) = union(terms(t.lhs), terms(t.rhs))
+terms(t::InteractionTerm) = terms(t.terms)
+terms(t::AbstractTerm) = [t]
+terms(t::NTuple{N, AbstractTerm}) where N = mapreduce(terms, union, t)
+
+needs_schema(t::AbstractTerm) = true
+needs_schema(::ConstantTerm) = false
+needs_schema(::FunctionTerm) = false
+needs_schema(t) = false
+
+schema(dt::D, hints=Dict{Symbol,Any}()) where {D<:Data.Table} =
+    schema(Term.(collect(fieldnames(D))), dt, hints)
+
+schema(data, hints=Dict{Symbol,Any}()) = schema(Data.stream!(data, Data.Table), hints)
+schema(ts::Vector{<:AbstractTerm}, data, hints::Dict{Symbol}) =
+    schema(Data.stream!(data, Data.Table), hints)
+
+# handle hints:
+function schema(ts::Vector{<:AbstractTerm}, dt::Data.Table, hints::Dict{Symbol})
+    sch = Dict{Any,Any}()
+    for t in ts
+        if t.sym ∈ keys(hints)
+            sch[t] = schema(t, dt, hints[t.sym])
+        else
+            sch[t] = schema(t, dt)
+        end
+    end
+    return sch
+end
+
+schema(f::FormulaTerm, data, hints::Dict{Symbol}) =
+    schema(filter(needs_schema, terms(f)), data, hints)
+
+schema(f::FormulaTerm, data) = schema(f, data, Dict{Symbol,Any}())
+
+schema(t::Term, dt::Data.Table) = schema(t, dt[t.sym])
+schema(t::Term, dt::Data.Table, hint) = schema(t, dt[t.sym], hint)
+
+schema(t::Term, xs::AbstractVector{<:Number}) = schema(t, xs, ContinuousTerm)
+schema(t::Term, xs::AbstractVector, ::Type{ContinuousTerm}) =
+    ContinuousTerm(t.sym, fit!(Series(Variance()), xs))
+
+# default contrasts: dummy coding
+schema(t::Term, xs::AbstractVector) = schema(t, xs, CategoricalTerm)
+schema(t::Term, xs::AbstractArray, ::Type{CategoricalTerm}) = schema(t, xs, DummyCoding())
+
+function schema(t::Term, xs::AbstractArray, contrasts::AbstractContrasts)
+    counts = fit!(Series(CountMap(eltype(xs))), xs)
+    contrmat = ContrastsMatrix(contrasts, collect(keys(counts.stats[1])))
+    CategoricalTerm(t.sym, counts, contrmat)
+end
+
+# TODO: add methods for schema(::Continuous/CategoricalTerm) (to re-set/validate schema)
+
+
+
+# now to _set_ the schema in formula...most straightforward way is to just look
+# up each term in the schema and replace it (calculating width of interaction
+# terms and things like that) but we want to handle the rank correcting stuff
+# too.  maybe that's best thought of as a kind of re-write rule?  or as a
+# special kind of schema/wrapper type?  then if it's just a dict, the "vanilla"
+# version is available...
+#
+# so what does that wrapper look like?  holds onto the terms that have been seen
+# so far, checks against that...
+
+apply_schema(ft::FormulaTerm, data::Data.Table, args...) =
+    apply_schema(ft, schema(ft, data, args...))
+
+apply_schema(t, schema) = t
+apply_schema(terms::NTuple{N,AbstractTerm}, schema) where N = apply_schema.(terms, Ref(schema))
```

Comment:
Same as the `N` type parameter on `CategoricalTerm`: so you can determine the width of the model matrix (or length of a single row) from the types of the terms alone.  Although it might make more sense to do that transformation RIGHT before we generate the row to keep the compilation time down...but that adds a level of indirection/complexity.  I suppose it would also be possible to add methods for `model_cols(::AbstractVector{<:AbstractTerm}, data)` that might just be slower (probably only noticeable when you're doing a lot of single-row `model_cols`).

Reply:
OK. I guess that makes sense if one can theoretically create a model with a formula and a predefined set of categorical levels so that it can completely be compiled statically.

Reply:
How "wide" do you expect things to ever get here? (i.e. how many tuple elements) Having tuples introduces cascading compilation and in really wide cases can be quite slow.

Reply:
I'd say less than 100, but often more than 10. The more I think about it the more it makes sense to keep from creating tuples until the very last stages (and even then maybe only use tuples in the case of row-based/streaming mode)